File: examples/go/intent-current-block-height/IntentCurrentBlockHeight.runar.go
package x

import runar "github.com/icellan/runar/packages/runar-go"

// IntentCurrentBlockHeight exercises the runar.CurrentBlockHeight
// shorthand, which is pure source-level sugar for
// `runar.ExtractLocktime(this.txPreimage)`. The desugar happens at ANF
// lowering time — no new ANF kind or stack codegen is needed.
type IntentCurrentBlockHeight struct {
	runar.StatefulSmartContract

	Deadline runar.Bigint `runar:"readonly"`
	Count    runar.Bigint
}

func (c *IntentCurrentBlockHeight) Spend() {
	h := runar.CurrentBlockHeight()
	runar.Assert(h <= c.Deadline)
	c.Count = c.Count + 1
}

File: examples/rust/intent-current-block-height/IntentCurrentBlockHeight.runar.rs
use runar::prelude::*;

/// IntentCurrentBlockHeight -- exercises the `current_block_height`
/// shorthand, which is pure source-level sugar for
/// `extract_locktime(self.tx_preimage)`.
#[runar::contract]
pub struct IntentCurrentBlockHeight {
    #[readonly]
    pub deadline: Bigint,
    pub count: Bigint,
}

impl IntentCurrentBlockHeight {
    pub fn spend(&mut self) {
        let h = current_block_height();
        assert!(h <= self.deadline);
        self.count = self.count + 1;
    }
}
